### DISCUSSION TITLE
Is POST /graphql/{query-name} a good idea?

### DISCUSSION THREAD

**Discussion by lburja** [graphql-java/graphql-java#2230] [Q&A] [answered]
When you expose a GraphQL Controller (using `graphql-java-spring-webmvc` or other frameworks), it generally registers an endpoint like `/graphql`.

Later on, when you inspect the requests using Chrome's Network tab, all requests look the same (`POST /graphql`). The same is true for the access logs, you see many identical lines (unless you also log the payload). This makes of course debugging more difficult, and renders the access logs less useful.

I was wondering if it's a good idea, to allow in my application `/graphql/*` style requests. For examples `/graphql/movies` would actually be equivalent to `/graphql` (and the `/movies` part is just the name of the query, and is for information purposes only).

Would this somehow be considered a bad idea, or would be for some reason undesirable? I couldn't really find anything in the spec for or against this technique.

**Comment by gklijs** [comment, marked as answer]
I don't really get what the advantage would be?

**Reply by lburja** [reply]
Well, consider the following lines in the access log:

```
10.1.254.228 - 2021-03-01 10:43:09,196 http-nio-8082-exec-4 "POST /api/graphql HTTP/1.1" 200 696 (74 ms)
10.1.254.228 - 2021-03-01 10:43:09,289 http-nio-8082-exec-2 "POST /api/graphql HTTP/1.1" 200 2481 (17599 ms)
10.1.254.228 - 2021-03-01 10:43:22,819 http-nio-8082-exec-3 "POST /api/graphql HTTP/1.1" 200 696 (76 ms)
```

I can't really tell what each of those `/api/graphql` queries is doing.  And one of them (the second one) is really slow. 
If on the other hand I had the following, it would be much more useful:

```
10.1.254.228 - 2021-03-01 10:43:09,196 http-nio-8082-exec-4 "POST /api/graphql/currentUser HTTP/1.1" 200 696 (74 ms)
10.1.254.228 - 2021-03-01 10:43:09,289 http-nio-8082-exec-2 "POST /api/graphql/meetings HTTP/1.1" 200 2481 (17599 ms)
10.1.254.228 - 2021-03-01 10:43:22,819 http-nio-8082-exec-3 "POST /api/graphql/documents HTTP/1.1" 200 696 (76 ms)
```

In this case, I can understand what's happening (e.g. I can tell that the `/api/graphql/meetings` query is slow). 

This is just an example. There are potentially other places where having a human readable URL might help, beyond access logs (e.g. developers looking in the Chrome Network tab, Server Monitoring Tools, etc.).

**Reply by gklijs** [reply]
Yes, but there is nu real correlation. And there are tools which do allow you to know which queries are being run and a lot more, like Apollo Studio https://www.apollographql.com/docs/studio/setup-analytics/. And it's confusing for people familiar with Graphql, and a lot of client's only allow one url + it would be a bit weird to send different queries to different urls, just to have them right in the access log.

**Reply by bbakerman** [reply]
The way to address this is to no use path as the key to access logs but the query

At Atlassian (where I work) We use the AST signature mechanism in graphql-java to make a canonical representation of the query

```
public class PrivacySafeQuery {

    /**
     * Makes the given query document PII and UGC safe.
     *
     * @param queryDoc      the query
     * @param operationName the name of the operation being executed
     * @return a compact dump of the executing query that has input data removed
     */
    public static String privacySafe(Document queryDoc, String operationName) {
        AstSignature astSignature = new AstSignature();
        Document safeDoc = astSignature.signatureQuery(queryDoc, operationName);
        return AstPrinter.printAstCompact(safeDoc);
    }

```

This gets put into access logs mainly for "security reasons" - we can reconstruct how did what later

We also do this for privacy reasons - we have the query but not the data that goes into the query - eg GDPR reasons

Finally we do use Apollo Studio to break down all this data which is first class (and costs at scale)

**Reply by bbakerman** [reply]
Another way to address this to elevate "operation" as the unit of tracking.

Before we used path as "the unit of tracking" that is for a path /api/foo/bar that names represents what consumers are after certain data.
in graphql operation cna represent this.  And operation called "getFoo" can represent the idea that peoeple are after certain data.

Operations may seem problematic because any one can call any query any name.  But in practice teams don;t.  So it can lead to insights on what is  being asked for.

Apollo Studio uses operation as the major key in their UI (albeit with smarts to allow for multiple queries to have the same operation name)

**Reply by lqc** [reply]
Using the query in contexts where you process it anyways, is definitely the best solution. On the other hand, there is also plenty of middleware that is HTTP-focused and it might not be possible/feasible to parse the query at all.

My solution to this is adding the "operationName" to the query string, e.g. `/api/graphql?operationName=FetchRecentMovies`. The operation name of course matches an actual query/mutation name inside the request so that everything conforms to the spec.

**Reply by lburja** [reply]
Thanks, this is useful! 

In the end I went with `POST /graphql?description=<graphql-operation-name>` .  

Adding the operation name as a query parameter makes it compatible with the current GraphQL controller, and naming it "description" is a good indication that the operation name is for information purposes only
